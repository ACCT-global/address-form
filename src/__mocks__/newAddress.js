export default {
  addressId: '1',
  addressType: 'residential',
  city: null,
  complement: null,
  country: null,
  geoCoordinates: null,
  neighborhood: null,
  number: null,
  postalCode: null,
  receiverName: null,
  reference: null,
  state: null,
  street: null,
}
