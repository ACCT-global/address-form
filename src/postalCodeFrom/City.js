import React, { Component } from 'react'
import PropTypes from 'prop-types'
import AddressShape from '../propTypes/AddressShape'
import find from 'lodash/find'
import map from 'lodash/map'

class City extends Component {
  handleStateChange = e => {
    const state = e.target.value

    this.props.onChangeAddress({
      ...this.props.address,
      state,
    })
  };

  handleCityChange = e => {
    const value = e.target.value

    this.props.onChangeAddress({
      ...this.props.address,
      ...this.deComposeValue(value),
    })
  };

  composeValue = address =>
    (address.city && address.postalCode
      ? `${address.city}___${address.postalCode}`
      : '');

  deComposeValue = value => {
    const [city, postalCode] = value.split('___')
    return { city, postalCode }
  };

  render() {
    const { address, rules } = this.props

    const stateOptions = find(rules.fields, field => field.field === 'state')
      .options

    return (
      <div>
        <label>
          State
          <select value={address.state || ''} onChange={this.handleStateChange}>
            <option value="" />
            {stateOptions.map(state => (
              <option key={state} value={state}>
                {state}
              </option>
            ))}
          </select>
        </label>
        <label>
          City
          <select
            value={this.composeValue(address)}
            onChange={this.handleCityChange}
          >
            {address.state
              ? map(rules.citiesPostalCodes[address.state], ({
                  postalCode,
                  city,
                }) => (
                  <option
                    key={city}
                    value={this.composeValue({ city, postalCode })}
                  >
                    {city}
                  </option>
                ))
              : null}
          </select>
        </label>
      </div>
    )
  }
}

City.propTypes = {
  address: PropTypes.shape(AddressShape),
  rules: PropTypes.object.isRequired,
  onChangeAddress: PropTypes.func.isRequired,
}

export default City
