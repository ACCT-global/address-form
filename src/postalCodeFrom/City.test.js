import React from 'react'
import renderer from 'react-test-renderer'
import City from './City'
import { shallow } from 'enzyme'
import CHL from '../country/CHL'
import newAddress from '../__mocks__/newAddress'

describe('City', () => {
  it('without state selected', () => {
    const address = {
      ...newAddress,
      country: 'CHL',
    }

    const tree = renderer
      .create(
        <City address={address} rules={CHL} onChangeAddress={jest.fn()} />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('with state selected', () => {
    const address = {
      ...newAddress,
      country: 'CHL',
      state: 'XV Región',
    }

    const tree = renderer
      .create(
        <City address={address} rules={CHL} onChangeAddress={jest.fn()} />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should change the state', () => {
    const handleChange = jest.fn()
    const wrapper = shallow(
      <City
        address={{
          ...newAddress,
          country: 'CHL',
        }}
        rules={CHL}
        onChangeAddress={handleChange}
      />
    )

    const event = { target: { value: 'XV Región' } }
    wrapper.find('select[name="state"]').simulate('change', event)

    expect(handleChange).toHaveBeenCalledWith({
      ...newAddress,
      country: 'CHL',
      state: 'XV Región',
    })
  })

  it('should change the city and postal code ', () => {
    const handleChange = jest.fn()
    const wrapper = shallow(
      <City
        address={{
          ...newAddress,
          country: 'CHL',
          state: 'XV Región',
        }}
        rules={CHL}
        onChangeAddress={handleChange}
      />
    )

    const event = { target: { value: 'Arica___1000000' } }
    wrapper.find('select[name="city"]').simulate('change', event)

    expect(handleChange).toHaveBeenCalledWith({
      ...newAddress,
      country: 'CHL',
      state: 'XV Región',
      city: 'Arica',
      postalCode: '1000000',
    })
  })
})
