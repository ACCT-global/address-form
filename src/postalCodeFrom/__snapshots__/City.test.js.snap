// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`City with state selected 1`] = `
<div>
  <label>
    State
    <select
      name="state"
      onChange={[Function]}
      value="XV Región"
    >
      <option
        value=""
      />
      <option
        value="Región Metropolitana"
      >
        Región Metropolitana
      </option>
      <option
        value="I Región"
      >
        I Región
      </option>
      <option
        value="II Región"
      >
        II Región
      </option>
      <option
        value="III Región"
      >
        III Región
      </option>
      <option
        value="IV Región"
      >
        IV Región
      </option>
      <option
        value="V Región"
      >
        V Región
      </option>
      <option
        value="VI Región"
      >
        VI Región
      </option>
      <option
        value="IX Región"
      >
        IX Región
      </option>
      <option
        value="VII Región"
      >
        VII Región
      </option>
      <option
        value="VIII Región"
      >
        VIII Región
      </option>
      <option
        value="X Región"
      >
        X Región
      </option>
      <option
        value="XI Región"
      >
        XI Región
      </option>
      <option
        value="XII Región"
      >
        XII Región
      </option>
      <option
        value="XIV Región"
      >
        XIV Región
      </option>
      <option
        value="XV Región"
      >
        XV Región
      </option>
    </select>
  </label>
  <label>
    City
    <select
      name="city"
      onChange={[Function]}
      value=""
    >
      <option
        value="Arica___1000000"
      >
        Arica
      </option>
      <option
        value="Camarones___1040000"
      >
        Camarones
      </option>
      <option
        value="General Lagos___1080000"
      >
        General Lagos
      </option>
      <option
        value="Putre___1070000"
      >
        Putre
      </option>
    </select>
  </label>
</div>
`;

exports[`City without state selected 1`] = `
<div>
  <label>
    State
    <select
      name="state"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      />
      <option
        value="Región Metropolitana"
      >
        Región Metropolitana
      </option>
      <option
        value="I Región"
      >
        I Región
      </option>
      <option
        value="II Región"
      >
        II Región
      </option>
      <option
        value="III Región"
      >
        III Región
      </option>
      <option
        value="IV Región"
      >
        IV Región
      </option>
      <option
        value="V Región"
      >
        V Región
      </option>
      <option
        value="VI Región"
      >
        VI Región
      </option>
      <option
        value="IX Región"
      >
        IX Región
      </option>
      <option
        value="VII Región"
      >
        VII Región
      </option>
      <option
        value="VIII Región"
      >
        VIII Región
      </option>
      <option
        value="X Región"
      >
        X Región
      </option>
      <option
        value="XI Región"
      >
        XI Región
      </option>
      <option
        value="XII Región"
      >
        XII Región
      </option>
      <option
        value="XIV Región"
      >
        XIV Región
      </option>
      <option
        value="XV Región"
      >
        XV Región
      </option>
    </select>
  </label>
  <label>
    City
    <select
      name="city"
      onChange={[Function]}
      value=""
    />
  </label>
</div>
`;
